datalad_dataverse/remote.py

```diff
@@ -92,111 +136,311 @@ def api(self):
 
     def checkpresent(self, key):
-        dataset = self.api.get_dataset(identifier=self.doi)
-
-        datafiles = dataset.json()['data']['latestVersion']['files']
-        if next((item for item in datafiles if item['dataFile']['filename'] == key), None):
-            return True
+        known_state = self.annex.getstate(key)
+        if known_state:  # empty string otherwise
+            return int(known_state) in self.files_all.keys()
         else:
-            return False
+            return Path(key) in [p for p in self.files_all.values()]
 
     def checkpresentexport(self, key, remote_file):
-        return self.checkpresent(key=remote_file)
-
-    def transfer_store(self, key, local_file, datafile=None):
-        ds_pid = self.doi
-        if datafile is None:
-            datafile = Datafile()
-            datafile.set({'filename': key, 'label': key})
-        datafile.set({'pid': ds_pid})
-        
-        resp = self.api.upload_datafile(identifier=ds_pid, filename=local_file, json_str=datafile.json())
-        resp.raise_for_status()
-
-    def transferexport_store(self, key, local_file, remote_file):
-        remote_dir = os.path.dirname(remote_file)
-        if re.search(pattern='[^a-z0-9_\-.\\/\ ]', string=remote_dir, flags=re.ASCII | re.IGNORECASE):
-            self.annex.error(f"Invalid character in directory name of {remote_file}."
-                             f"Valid characters are a-Z, 0-9, '_', '-', '.', '\\', '/' and ' ' (white space).")
-
-        datafile = Datafile()
-        datafile.set({'filename': remote_file,
-                      'directoryLabel': remote_dir,
-                      'label': os.path.basename(remote_file)})
-        self.transfer_store(key=remote_file, local_file=local_file, datafile=datafile)
 
-    def transfer_retrieve(self, key, file):
+        # TODO: Properly test whether presentexport should consider all dataset
+        # versions or just latest. Problem: To retrieve "old" keys, all versions
+        # would be relevant, but for upload a reexport of the same file may be
+        # the correct thing (present in version 1, removed in 2, about to be
+        # present in 3 again).
+        known_state = self.annex.getstate(key)
+        if known_state:  # empty string otherwise
+            return int(known_state) in self.files_latest.keys()
+        else:
+            return mangle_directory_names(remote_file) in \
+                   [p for p in self.files_latest.values()]
+
+    def _upload_file(self, datafile, key, local_file, remote_file):
+        """helper for both transfer-store methods"""
+        # If the remote path already exists, we need to replace rather than
+        # upload the file, since otherwise dataverse would rename the file on
+        # its end. However, this only concerns the latest version of the dataset
+        # (which is what we are pushing into)!
+        existing_id = [i for i, p in self.files_latest.items()
+                       if p == remote_file]
```

Comment:
OK, more complexity I was not aware of until now. The dataverse file ID do not seem to be content-specific.

That makes me wonder how this all could possibly work...

If we initially upload a file and it get's ID=1. Then we release the dataset. Now we upload new file content to a new version of the dataset. The code below explicitly replaces that file (by ID). What happens to the released dataset version? It likely did not replace the file content for a previous version.